core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala

```diff
@@ -470,32 +450,85 @@ class SpendReportingService(
     bytesProcessedCounter += stats.getEstimatedBytesProcessed
   }
 
+  def getSpendReportableWorkspaceGoogleProjects(
+    childContext: RawlsRequestContext
+  ): Future[Map[RawlsBillingProjectName, Seq[Workspace]]] =
+    samDAO
+      .listResourcesWithActions(
+        SamResourceTypeNames.workspace,
+        SamWorkspaceActions.readSpendReport,
+        childContext
+      )
+      .flatMap {
+        case ownerWorkspaces if ownerWorkspaces.isEmpty =>
+          Future.successful(Map.empty[RawlsBillingProjectName, Seq[Workspace]])
+        case ownerWorkspaces =>
+          val validWorkspaceIds = ownerWorkspaces
+            .map(_.getResourceId)
+            .filter(resourceId => Try(UUID.fromString(resourceId)).isSuccess) // filter out non-UUIDs
+            .toList
+          workspaceServiceConstructor(childContext).getGCPWorkspacesByBillingProjects(validWorkspaceIds)
+      }
+
+  def getSpendReportableWorkspaceGoogleProjectsInBillingProject(
+    projectName: RawlsBillingProjectName,
+    childContext: RawlsRequestContext
+  ): Future[Map[GoogleProjectId, WorkspaceName]] = {
+
+    def filterProjects(projects: Map[RawlsBillingProjectName, Seq[Workspace]]): Map[GoogleProjectId, WorkspaceName] =
+      projects.flatMap { case (_, value) =>
+        value.collect {
+          case workspace if workspace.namespace == projectName.value =>
+            workspace.googleProjectId -> workspace.toWorkspaceName
+        }
+      }
+
+    for {
+      projects <- getSpendReportableWorkspaceGoogleProjects(childContext)
+      filteredProjects = filterProjects(projects)
+      hasAction <- samDAO.userHasAction(SamResourceTypeNames.billingProject,
+                                        projectName.value,
+                                        SamBillingProjectActions.readSpendReport,
+                                        childContext
+      )
+      result <-
+        if (hasAction || filteredProjects.nonEmpty) {
+          Future.successful(filteredProjects)
+        } else {
+          Future.failed(
+            RawlsExceptionWithErrorReport(
+              StatusCodes.Forbidden,
+              s"${childContext.userInfo.userEmail.value} cannot perform ${SamBillingProjectActions.readSpendReport.value} on project ${projectName.value}"
+            )
+          )
+        }
+    } yield result
+  }
```

Comment:
This is far too complicated. Checking the billing project action is not required, access at the workspace level has already been checked and that is sufficient. The filter function does not need to cycle through all the workspaces because the map is already keyed by billing project.

This function should be 
```
getSpendReportableWorkspaceGoogleProjects(childContext).map { workspacesByProject =>
  workspacesByProject.get(projectName).getOrElse(Map.empty).map { workspace =>
    workspace.googleProjectId -> workspace.toWorkspaceName
  }.toMap
}
```

Reply:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/6ca541dbe6e14fd161620ceb45f298cda6aabc57